node/sync/src/block_sync.rs

```diff
@@ -215,6 +215,13 @@ impl<N: Network> BlockSync<N> {
         let block_requests = self.prepare_block_requests();
         trace!("Prepared {} block requests", block_requests.len());
 
+        // If there are no block requests, then try if we can advance with responses we already
+        // have in the sync pool.
+        if block_requests.is_empty() && !self.responses.read().is_empty() {
+            trace!("No block requests to send - try handling responses we already have");
+            self.handle_responses(self.canon.latest_block_height());
```

Comment:
This is not a fix of the issue. There is `construct_request()` that calculates `num_sync_ips` which is sometimes `REDUNDANCY_FACTOR=3`.

The fix should be done in other way:

```diff
/// A struct that keeps track of the current block sync state.
///
/// # State
/// - When a request is inserted, the `requests` map and `request_timestamps` map insert an entry for the request height.
/// - When a response is inserted, the `requests` map inserts the entry for the request height.
-/// - When a request is completed, the `requests` map still has the entry, but its `sync_ips` is empty;
+/// - When a request is completed, the `requests` map still has the entry, but its `sync_ips` is empty AND num_sync_ips = 0;
///   the `request_timestamps` map remains unchanged.
/// - When a response is removed/completed, the `requests` map and `request_timestamps` map also remove the entry for the request height.
/// - When a request is timed out, the `requests`, `request_timestamps`, and `responses` map remove the entry for the request height;
#[derive(Clone, Debug)]
pub struct BlockSync<N: Network> {
...

/// Inserts the given block response, after checking that the request exists and the response is well-formed.
-/// On success, this function removes the peer IP from the requests map.
+/// On success, this function removes the peer IP from the requests map AND decrements num_sync_ips.
/// On failure, this function removes all block requests from the given peer IP.
fn insert_block_response(&self, peer_ip: SocketAddr, block: Block<N>) -> Result<()> { ... }


fn remove_timed_out_block_requests(&self) -> usize {
...
    // Determine if the request is incomplete.
    let is_request_incomplete =
-        !requests.get(height).map(|(_, _, peer_ips)| peer_ips.is_empty()).unwrap_or(false);
+        !requests.get(height).map(|(_, _, _, num_sync_ips)| num_sync_ips > 0).unwrap_or(false);
```

Reply:
This doesn't even compile.

Reply:
@kpp, we've taken another look at the logic, I'll attempt to summarise here. The checks implemented on snarkOS (and down in VM) when progressing to the next block should be sufficient to allay the concerns you've mentioned above about exploiting timeouts in sync to misguide a validator. Specifically, we check the authority of the block (amongst other things)—this makes receiving even a single block safe for sync purposes. In short, we don't rely on communication redundancy for safety. Don't hesitate to follow up if you're unsure about this and still have concerns. 

Please note that sync logic is in flux at the moment, this PR is part of a larger effort to improve this aspect of the network.

Reply:
Thank you!